src/core/lib/gprpp/time.cc

```diff
@@ -133,6 +142,14 @@ int64_t TimespanToMillisRoundDown(gpr_timespec ts) {
 
 }  // namespace
 
+GPR_THREAD_LOCAL(Timestamp::Source*)
+Timestamp::source_{NoDestructSingleton<GprNowTimeSource>::Get()};
+
+Timestamp ScopedTimeCache::Now() {
+  if (!cached_time_.has_value()) cached_time_ = previous()->Now();
```

Comment:
Instead of defaulting to the previous version of now, would it make sense to instead grab the current time using `gpr_now()` and then call `previous()->InvalidateCache()`, so that when the nested instance is popped off the stack, we don't jump backwards in time?

Reply:
`previous()` here is the previous time source. By always using it we'll be able to do some fancy stuff later (like implanting custom time mocks in scoped segments of code).

Reply:
My concern is that this seems like a behavior change.  Currently, if a nested `ExecCtx` is created, then `Now()` starts returning a new time (the one from the nested `ExecCtx` instance), not the one from the previous `ExecCtx` instance.  But with this change, we're no longer doing that, so creating a new nested `ExecCtx` no longer increases the current time, whereas I think it did so before this PR.

I do realize that this is partially a bug fix, because we don't want the time to go backwards when the nested `ExecCtx` is destroyed and we revert to the previous `ExecCtx`.  But that's why I suggested using the new time and then invalidating the time in the previous time source: that way, we will still increment the time when we create the nested `ExecCtx`, but we'll also avoid the time going backwards when the nested `ExecCtx` is destroyed.

As an alternative, how about simply calling `InvalidateNow()` on the previous time source when we instantiate a nested time source?  That way, we can get the same behavior while still depending on the previous instance for the value of now.

Reply:
I can live with that, done.